beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,264 @@
+    reweight["states"] = beluga::views::states(record.reweight) | ranges::to<std::vector<double>>;
+    reweight["weights"] = beluga::views::weights(record.reweight) | ranges::to<std::vector<double>>;
+
+    auto resample = node["simulation_records"][cycle]["particles"]["resample"];
+    resample["states"] = beluga::views::states(record.resample) | ranges::to<std::vector<double>>;
+    resample["weights"] = beluga::views::weights(record.resample) | ranges::to<std::vector<double>>;
+
+    auto estimation = node["simulation_records"][cycle]["estimation"];
+    estimation["mean"] = std::get<0>(record.estimation);
+    estimation["sd"] = std::get<1>(record.estimation);
+  }
+
+  std::ofstream fout(path, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+int main(int argc, char* argv[]) {
+  if (argc < 2) {
+    std::cerr << "Usage: " << argv[0] << "<parameters_path>\n";
+    return 1;
+  }
+
+  Parameters parameters;
+  load_parameters(argv[1], parameters);
+
+  SimulationRecord simulation_record;
+
+  LandmarkMapVector landmark_map;
+  load_landmark_map(argv[1], landmark_map);
+  simulation_record.landmark_map = landmark_map;
+
+  std::normal_distribution<double> initial_distribution(parameters.initial_position, parameters.initial_position_sd);
+  auto particles = beluga::views::sample(initial_distribution) |                  //
+                   ranges::views::transform(beluga::make_from_state<Particle>) |  //
+                   ranges::views::take_exactly(parameters.number_of_particles) |  //
+                   ranges::to<beluga::TupleVector>;
+
+  double current_position{parameters.initial_position};
+  for (unsigned int n = 0; n < parameters.number_of_cycles; ++n) {
+    RobotRecord robot_record;
+
+    current_position += parameters.velocity * parameters.dt;
+    robot_record.ground_truth = current_position;
+
+    if (current_position > parameters.map_size) {
+      break;
+    }
+
+    auto motion_model = [&](double particle_position, auto& gen) {
+      const double distance = parameters.velocity * parameters.dt;
+      std::normal_distribution<double> distribution(distance, parameters.translation_sd);
+      return particle_position + distribution(gen);
+    };
+
+    auto detections =
+        landmark_map |                                                                                       //
+        ranges::views::transform([&](double landmark) { return landmark - current_position; }) |             //
+        ranges::views::remove_if([&](double range) { return std::abs(range) > parameters.sensor_range; }) |  //
+        ranges::to<SensorData>;
+
+    auto sensor_model = [&](double particle_position) {
+      auto particle_detections =
+          landmark_map |                                                                             //
+          ranges::views::transform([&](double landmark) { return landmark - particle_position; }) |  //
+          ranges::to<SensorData>;
```

Comment:
meta: The use of these aliases seems a bit inconsistent. With good variable names, I wonder if we need them.

```suggestion
          ranges::to<std::vector>;
```

In production code, they would have their place, but for a tutorial I wonder if we should value locality of information more (the reader doesn't have to find out what exactly `SensorData` is).

From the aliases defined above, I would only keep `Particle`.

Feel free to ignore this comment anyway.

Reply:
It makes sense. I started with this because, in the beginning, I used an std::vector<int> for the landmark_map. However, after I moved all the tutorial to a continuous space, I can erase them and leave only the definition of Particles.